### PR TITLE
[BUGFIX] [ENHANCEMENT] Allows gzipped files and other encodings to be read from S3

### PR DESCRIPTION
**Pull request by luke321321**
Changes proposed in this pull request:
- [BUG] Allow gzipped files (and other encodings) to be read from S3 using pandas.  Instead of decoding directly to `StringIO` and passing to pandas with kwargs we now let pandas to do the decoding implicitly for us by keeping as `BytesIO`.  This allows the pandas kwarg `compression` to work, which it wouldn't previously.  Not sure to classify this as a bug or enhancement.  
- [ENHANCEMENT] Attempt to detect compression from S3 key suffix (added features to utils to achieve this)
- Refactored some S3 pandas tests & added test for the above
- Removed redundant S3 pandas test - the test was a duplicate (this can be added back or fixed but I'm unsure how this part of the GE API works.  Should be quick for someone familiar to correct this test & add back in now)

Note for some reason using the `pd.testing.assert_frame_equals` wasn't working for me and I'm not sure why.  I've just left the test and checking the shapes are equal but this can be improved some what.

### CODE REVIEW

**Comment by eugmandel** [comment]
This looks like a great enhancement, @luke321321 ! Adding the review queue - we will get to this within a couple of days.

**Comment by luke321321** [comment]
Note happy to refactor `sniff_s3_compression()` to use a dictionary instead like:
```python
def sniff_s3_compression(s3_url: S3Url) -> str:
    """Attempts to get read_csv compression from s3_url"""
    suffix = s3_url.suffix
    suffix_to_kwarg = {"gz" : "gzip", ... }
    return suffix_to_kwarg.get(suffix, "infer")
```

**Comment by luke321321** [comment]
i've actually refactored for the above as I think it's easier to read, test & extend.  Functionality is identical.

**Comment by luke321321** [comment]
☝️  note test failure comes from a different part of the repo from a flakey test (`test_send_malformed_data`)

File: tests/execution_engine/test_pandas_execution_engine.py [hunk 1]
```diff
@@ -9,11 +10,6 @@
 from moto import mock_s3
```
Comment (roblim) [inline tests/execution_engine/test_pandas_execution_engine.py:12]:
```suggestion
from botocore.errorfactory import ClientError
from moto import mock_s3

```

File: tests/execution_engine/test_pandas_execution_engine.py [hunk 2]
```diff
@@ -9,11 +10,6 @@
...
 
 import great_expectations.exceptions.exceptions as ge_exceptions
```
Comment (roblim) [inline tests/execution_engine/test_pandas_execution_engine.py:13]:
```suggestion
import great_expectations.exceptions.exceptions as ge_exceptions
from great_expectations.core.batch import BatchDefinition
```

File: tests/execution_engine/test_pandas_execution_engine.py [hunk 3]
```diff
@@ -9,11 +10,6 @@
...
-from great_expectations.core.batch import Batch
-from great_expectations.datasource.data_connector import (
-    ConfiguredAssetS3DataConnector,
-    InferredAssetS3DataConnector,
-)
 from great_expectations.datasource.types.batch_spec import (
```
Comment (roblim) [inline tests/execution_engine/test_pandas_execution_engine.py]:
```suggestion
from great_expectations.core.id_dict import PartitionDefinition
from great_expectations.datasource.types.batch_spec import (
```

File: tests/execution_engine/test_pandas_execution_engine.py [hunk 4]
```diff
@@ -299,92 +295,119 @@ def test_get_batch_with_split_on_whole_table_filesystem(
...
-    client = boto3.client("s3", region_name=region_name)
 
-    test_df: pd.DataFrame = pd.DataFrame(data={"col1": [1, 2], "col2": [3, 4]})
+@pytest.fixture
+def test_s3_files(s3, s3_bucket, test_df_small_csv):
     keys: List[str] = [
         "path/A-100.csv",
         "path/A-101.csv",
         "directory/B-1.csv",
         "directory/B-2.csv",
```
Comment (roblim) [inline tests/execution_engine/test_pandas_execution_engine.py:349]:
```suggestion
        "directory/B-2.csv",
        "alpha-1.csv",
        "alpha-2.csv",
```

File: tests/execution_engine/test_pandas_execution_engine.py [hunk 5]
```diff
@@ -299,92 +295,119 @@ def test_get_batch_with_split_on_whole_table_filesystem(
...
-            batch_spec=S3BatchSpec(
-                path=full_path,
-                reader_method="read_csv",
-                splitter_method="_split_on_whole_table",
-            )
+            batch_spec=batch_with_split_on_whole_table_s3
+        )
+
+
+@pytest.fixture
```
Comment (roblim) [inline tests/execution_engine/test_pandas_execution_engine.py]:
```suggestion
            batch_spec=batch_with_split_on_whole_table_s3
        )


def test_get_batch_with_split_on_whole_table_s3_with_configured_asset_s3_data_connector(test_s3_files, test_df_small):
    bucket, _keys = test_s3_files
    expected_df = test_df_small

    my_data_connector = ConfiguredAssetS3DataConnector(
        name="my_data_connector",
        datasource_name="FAKE_DATASOURCE_NAME",
        default_regex={
            "pattern": "alpha-(.*)\\.csv",
            "group_names": ["index"],
        },
        bucket=bucket,
        prefix="",
        assets={"alpha": {}},
    )
    batch_def = BatchDefinition(
        datasource_name="FAKE_DATASOURCE_NAME",
        data_asset_name="alpha",
        data_connector_name="my_data_connector",
        partition_definition=PartitionDefinition(index=1),
        batch_spec_passthrough={
            "reader_method": "read_csv",
            "splitter_method": "_split_on_whole_table",
        },
    )
    test_df = PandasExecutionEngine().get_batch_data(
        batch_spec=my_data_connector.build_batch_spec(batch_definition=batch_def)
    )
    assert test_df.dataframe.shape == expected_df.shape

    # if key does not exist
    batch_def_no_key = BatchDefinition(
        datasource_name="FAKE_DATASOURCE_NAME",
        data_asset_name="alpha",
        data_connector_name="my_data_connector",
        partition_definition=PartitionDefinition(index=9),
        batch_spec_passthrough={
            "reader_method": "read_csv",
            "splitter_method": "_split_on_whole_table",
        },
    )
    with pytest.raises(ClientError):
        PandasExecutionEngine().get_batch_data(
            batch_spec=my_data_connector.build_batch_spec(
                batch_definition=batch_def_no_key
            )
        )


@pytest.fixture
```

**Review by roblim — CHANGES_REQUESTED** [review]
Looks awesome @luke321321! Turns out the duplicate test was meant to be filled out but wasn't - great catch! I went ahead and added the completed test back inline with my review. 

Would you mind just taking care of the new merge conflicts? Once the checks are passing again, we can approve/merge. Thanks so much!

**Comment by luke321321** [comment]
Hi @roblim thanks for adding that test back in.  I've fixed the merge conflicts.

**Review by roblim — APPROVED** [review]
Thanks @luke321321 - looks great!

**Pull request merged**